### PR TITLE
fix(block-no-verify): treat a git command line quoted as data as data, not a command

### LINKED ISSUE #3023: block-no-verify: false positive when a git command appears inside a quoted argument to another program
## Summary

`scripts/hooks/block-no-verify.js` blocks a Bash command when a git bypass flag appears **inside a quoted argument to a different program**, even though no git command is being run. The guard's own tokenizer and `VALUE_OPTIONS` work already handles the commit-message and `echo` cases correctly — this is the remaining gap.

Version: **ECC 2.2.1** (`ecc@ecc`), macOS, Claude Code.

## Reproducer

Feed the hook a `PreToolUse` payload whose `tool_input.command` is:

```
node /tmp/cli.js 'git commit --no-verify -m x'
```

Result: **BLOCKED** — `BLOCKED: --no-verify flag is not allowed with git commit.`

Nothing here runs git; the quoted string is `argv[2]` of `node`. The same happens when the string is piped rather than passed as an argument:

```
printf '%s' 'git commit --no-verify -m x' | node /tmp/x.js
```

Result: **BLOCKED**.

## What already works (so this is a narrow edge, not a regression)

These are all correctly **allowed** on 2.2.1, so the `VALUE_OPTIONS` / `tokenizeShellWords` work is doing its job:

| command | verdict |
|---|---|
| `git commit -m "document that --no-verify is banned"` | allowed |
| `echo "never use --no-verify with git commit"` | allowed |
| `python3 -c "print('docs mention git commit and --no-verify')"` | allowed |
| `git commit -m "safe message"` | allowed |

And the true positives still block, as they must:

| command | verdict |
|---|---|
| `git commit --no-verify -m "real bypass"` | BLOCKED |
| `git push --no-verify` | BLOCKED |

## Why it matters in practice

I hit this twice in one session doing legitimate work:

1. Running a CLI **under test** with a command string as its argument — exactly the reproducer above.
2. Writing a note documenting the guard's own behaviour: the prose contained the flag, so an unrelated `python3` heredoc was refused.

The failure mode is awkward because the guard is hardest to work on, document, or test precisely when you are working on, documenting, or testing it. The workaround I used was `git commit -F <file>` plus assembling the token from parts, which is not discoverable.

## Suggested direction

After tokenizing, treat `git` as a command only when it is in **argv0 position of a statement** — start of input, or immediately after a control operator (`;`, `&&`, `||`, `|`). A `git` inside a single- or double-quoted token that is an *argument* to some other argv0 is data, not a command.

For reference, the upstream `tupe12334/block-no-verify` package hit the same class of problem (their #70, a `-n` belonging to a chained command). On their current `main`, `checkCommand()` allows all four cases in the first table and still blocks both in the second — ECC's local implementation already matches that; this quoted-argument case looks like the one remaining edge.

Happy to test a patch against the same case list.

### PR DESCRIPTION
**Pull request by L4XB**
## What Changed
`scripts/hooks/block-no-verify.js`: when `findGit()` finds a `git` inside a quoted string, the new `isQuotedDataArgument()` looks at the argv0 of the statement that carries the quote (first word after the last top-level `;`, `|`, `&` or newline, skipping `VAR=value` assignments). If that argv0 is a shell or command wrapper (`sh`, `bash`, `zsh`, `dash`, `ksh`, `fish`, `busybox`, `eval`, `exec`, `command`, `xargs`, `sudo`, `doas`, `su`, `env`, `nice`, `nohup`, `timeout`, `time`, `watch`, `flock`, `ssh`, `script`) or `git` itself, the quoted git is still inspected as before. For any other program the quoted string is data and the occurrence is skipped. Unquoted `git` anywhere in a chain is unaffected.

Tests added to `tests/hooks/block-no-verify.test.js`:
- allow: `node /tmp/cli.js 'git commit --no-verify -m x'`, `node /tmp/cli.js "git push --no-verify"`, `printf '%s' 'git commit --no-verify -m x' | node /tmp/x.js`, `python3 -c "print('git commit --no-verify')"`, `grep -n 'git commit --no-verify' docs/hooks.md`
- block: `sh -c "git commit --no-verify -m x"`, `bash -lc 'git push --no-verify'`, `sudo git commit --no-verify -m x`, `xargs -0 git commit --no-verify`, `env FOO=1 git commit -n -m x`, `eval 'git commit --no-verify -m x'`, `node /tmp/cli.js "data" && git commit --no-verify -m x`, `echo 'git commit --no-verify' ; git push --no-verify`

## Why This Change
Fixes #3023. `findGit()` accepted a `git` preceded by a quote character wherever it appeared, so `node /tmp/cli.js 'git commit --no-verify -m x'` (a CLI under test) and `printf '%s' 'git commit --no-verify -m x' | node /tmp/x.js` were blocked although nothing runs git. That made the guard hardest to work on, document or test exactly when doing so. The quote-before-git acceptance is still needed for `sh -c "git ..."`, so the fix keys on who receives the quoted string rather than dropping quoted matches altogether.

One case changes from allow to block as a side effect: `echo 'git commit --no-verify' ; git push --no-verify` was previously allowed, because the quoted `git commit` was inspected first, the stray quote derailed `findCommandSegmentEnd`, and the real `git push --no-verify` in the next segment was never reached. With the quoted occurrence skipped, the real one is found.

## Testing Done
- [x] Manual testing completed (the two reproducers from the issue plus the tables in the report)
- [x] Automated tests pass locally (`node tests/hooks/block-no-verify.test.js`: 31 passed, `node tests/hooks/cursor-block-no-verify.test.js`: 14 passed; the two new tests fail on `main`)
- [x] Edge cases considered and tested (wrappers, `VAR=value` prefix, chains after a quoted data argument, pipes)

## Type of Change
- [x] `fix:` Bug fix

## Security & Quality Checklist
- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly (none changed)
- [x] Shell scripts pass shellcheck (if applicable) (none changed)
- [x] Pre-commit hooks pass locally (if configured) (`npx eslint` on the changed files)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

### CODE REVIEW

File: scripts/hooks/block-no-verify.js [hunk 1]
```diff
@@ -35,6 +35,39 @@ const GIT_COMMANDS_WITH_NO_VERIFY = [
...
+  'env',
+  'nice',
+  'nohup',
+  'timeout',
+  'time',
+  'watch',
+  'flock',
+  'ssh',
+  'script',
+]);
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js:84]: Added in e1b13bc: `setsid`, `stdbuf`, `taskset`, `chrt`, `ionice`, `unshare`, `chroot`, `runuser`, `npx`/`bunx`/`pnpx` (plus `csh`/`tcsh` from the Greptile finding). Unknown argv0 with `-c` stays data. Tests: `blocks a quoted git command line behind a process prefix or another shell`.

File: scripts/hooks/block-no-verify.js [hunk 2]
```diff
@@ -284,6 +317,84 @@ function isInComment(input, idx) {
...
+ * A `git` inside a quoted string is only a command when that string is
+ * handed to something that executes it. Otherwise it is an argument of an
+ * unrelated program (a CLI under test, printf, python -c, ...) and must not
+ * be inspected for bypass flags.
+ */
+function isQuotedDataArgument(input, idx) {
+  const argv0 = quotedArgumentArgv0(input, idx);
+  if (argv0 === null || argv0 === '') return false;
+  const base = commandBasename(argv0);
+  return base !== 'git' && !COMMAND_WRAPPERS.has(base);
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js:362]: Done in e1b13bc. The quoted-region scan now records whether a double-quoted region contains `$(` or a backtick, and such a region is never treated as data. While adding the tests I found that `--no-verify)` / `--no-verify\`` were tokenized as one word (the closing delimiter attached to the flag), which hid the flag even outside quotes; unquoted `` ` ``, `(` and `)` now end a shell word. Tests: `echo "$(git commit --no-verify -m x)"`, `echo "$(git push --no-verify)"`, `printf "%s" "\`git push --no-verify\`"`.

File: scripts/hooks/block-no-verify.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js:84]: Added `csh` and `tcsh` (and the process prefixes from the CodeRabbit finding) to the wrapper set in e1b13bc, with regression tests for `csh -c "git commit --no-verify -m x"` and `tcsh -c 'git push --no-verify'`.

File: scripts/hooks/block-no-verify.js [hunk 3]
```diff
@@ -302,7 +413,9 @@ function findGit(input, start) {
     }
 
     const before = idx > 0 ? input[idx - 1] : ' ';
-    if (VALID_BEFORE_GIT.includes(before)) return { idx, len };
+    if (VALID_BEFORE_GIT.includes(before) && !isQuotedDataArgument(input, idx)) {
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js:383]: Fixed in e1b13bc: the quoted regions of the input are now collected in one forward pass (cached per input) and each `git` token is looked up with a binary search, so the hook is linear in the input size. New test `stays fast on a large quoted data payload full of git tokens` runs a 300 KB quoted payload (12k `git commit --no-verify` tokens) through the real hook and asserts exit 0 within the budget; it takes well under a second locally. Also restored `yarn.lock`, which had been modified by accident in the first commit.

File: scripts/hooks/block-no-verify.js [hunk 4]
```diff
@@ -284,6 +334,152 @@ function isInComment(input, idx) {
...
+    }
+
+    if (quote) {
+      if (quote === '"' && char === '\\') {
+        escaped = true;
+        continue;
+      }
+      if (char === quote) {
+        region.end = i;
+        region = null;
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Reworked: the scanner now keeps only local mutable state (`open`, `suspended`) and appends an immutable snapshot (`{ ...open, end }`) once a region is complete; nothing in `regions` is mutated afterwards.

File: scripts/hooks/block-no-verify.js [hunk 5]
```diff
@@ -284,6 +334,152 @@ function isInComment(input, idx) {
...
+  let region = null;
+  let argv0 = null;
+  let word = '';
+  let inWord = false;
+
+  const endWord = () => {
+    if (!inWord) return;
+    if (argv0 === null && word !== '' && !/^[A-Za-z_][A-Za-z0-9_]*=/.test(word)) {
+      argv0 = word;
+    }
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Fixed: shell reserved words (`if`/`then`/`do`/`{`/`}`/`!`/...) are skipped when picking a statement's argv0, and unquoted `(`, `)` and backticks start a new statement, so `( sh -c 'git commit --no-verify -m x' )`, `{ sh -c ...; }`, `if sh -c ...; then` and `for ...; do sh -c ...; done` are classified by the real command. While adding those tests I found that a quoted git command followed by more text in the same statement (`sh -c 'git push --no-verify'; echo done`) passed on `main` too: the flag scan started inside the string and treated its closing quote as an opener, gluing `; echo done` onto the flag token. The scan is now bounded by the quoted region the `git` sits in. All of these are regression tests now.

File: scripts/hooks/block-no-verify.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Fixed: a `$(` or backtick inside a double-quoted string now suspends that string; the substitution body is scanned as top-level shell (its own statements and quotes, nested substitutions included) and the string resumes as a new region after the matching `)` / backtick. `echo "$(echo "x"; git commit --no-verify -m x)"`, `echo "before $(sh -c 'git commit --no-verify -m x') after"` and `echo "$(echo "$(git push --no-verify)")"` are blocked and covered by tests.

File: scripts/hooks/block-no-verify.js [hunk 6]
```diff
@@ -284,6 +334,192 @@ function isInComment(input, idx) {
...
+      continue;
+    }
+
+    const outer = suspended.length > 0 ? suspended[suspended.length - 1] : null;
+    if (outer !== null) {
+      const resumes = outer.backtick ? char === '`' : char === ')' && outer.depth === 0;
+      if (resumes) {
+        suspended.pop();
+        newStatement();
+        openRegion(i, '"', outer.argv0, true);
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Fixed: each suspended entry now parks the outer `word` / `inWord` / `argv0`, the substitution body starts as a fresh statement, and the outer state is restored when the string resumes, so `FOO="pre$(echo x)post"` stays one assignment word and `sh` becomes argv0. That input and a backtick variant (`PREFIX="`date`" sh -c 'git push --no-verify'`) are blocked regression tests now.

File: scripts/hooks/block-no-verify.js [hunk 7]
```diff
@@ -284,6 +334,188 @@ function isInComment(input, idx) {
...
+    if (quote) {
+      if (quote === '"' && char === '\\') {
+        escaped = true;
+        continue;
+      }
+      if (char === quote) {
+        closeRegion(i, false);
+        continue;
+      }
+      if (quote === '"' && (char === '`' || (char === '
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: That review ran on 72d5dd9, where a `String.replace` replacement pattern (`$'`) had mangled the hunk; 622b085 restored the file (the test suite and `node --check` pass since). The scanner has now also moved out of this file, see below.

File: scripts/hooks/block-no-verify.js [hunk 8]
```diff
@@ -284,6 +334,188 @@ function isInComment(input, idx) {
...
+ * command substitution (`"$(git ...)"`, "`git ...`") runs git before any
+ * program receives it, so it is never data.
+ */
+function isQuotedDataArgument(input, idx) {
+  const region = quotedRegionAt(input, idx);
+  if (region === null || region.argv0 === '') return false;
+  if (region.substitution) return false;
+  const base = commandBasename(region.argv0);
+  return base !== 'git' && !COMMAND_WRAPPERS.has(base);
+}
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Done in the latest commit: the quote scanner lives in `scripts/lib/shell-quotes.js` (210 lines, small state-machine helpers) and the hook only imports `quotedRegionAt`. `block-no-verify.js` is 650 lines again; behaviour and tests unchanged.

File: scripts/hooks/block-no-verify.js [hunk 9]
```diff
@@ -284,6 +334,204 @@ function isInComment(input, idx) {
...
+      continue;
+    }
+
+    word += char;
+    inWord = true;
+  }
+
+  if (open !== null) regions.push({ ...open, end: input.length });
+  quotedRegionCache = { input, regions };
+  return regions;
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Split as requested: the scanner is now `scripts/lib/shell-quotes.js`, made of helpers under 50 lines each (`openRegion`, `closeRegion`, `suspendString`, `resumeString`, `scanQuotedChar`, `scanBareChar`, `quotedRegions`, `quotedRegionAt`), and `block-no-verify.js` is back to 650 lines.

File: scripts/hooks/block-no-verify.js [hunk 10]
```diff
@@ -289,6 +341,32 @@ function isInComment(input, idx) {
...
+ * command substitution (`"$(git ...)"`, "`git ...`") runs git before any
+ * program receives it, so it is never data.
+ */
+function isQuotedDataArgument(input, idx) {
+  const region = quotedRegionAt(input, idx);
+  if (region === null || region.argv0 === '') return false;
+  if (region.substitution) return false;
+  const base = commandBasename(region.argv0);
+  return base !== 'git' && !COMMAND_WRAPPERS.has(base);
+}
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js:417]: Valid, and fixed in f409441. Two separate holes were in play.

**1. A runtime given an eval flag runs its quoted argument as source.** The classification now asks whether the words between the program and the quote turn that argument into code, instead of looking at argv0 alone:

```js
const CODE_EVALUATORS = new Set(['node', 'bun', 'deno', 'python3', 'perl', 'ruby', 'php', 'lua', 'rscript', 'osascript', ...]);
const EVAL_FLAG = /(?:^|\s)(?:-{1,2}(?:e|eval|c|command|p|print|r)|eval)(?:=|\s|$)/i;
```

The quoted-region scanner records where the statement's argv0 starts, so the hook can read that span. `node -e`, `node --eval=`, `node -p`, `python3 -c`, `perl -e`, `ruby -e`, `php -r` and `deno eval` are scanned again; `node cli.js 'git commit --no-verify'`, `python3 /tmp/check.py '...'`, `printf '%s' '...'` and `grep -n '...' docs.md` stay data, which is what this PR set out to fix.

**2. The flag token carried the punctuation that closes the call.** `execSync("git push --no-verify")` tokenized to `--no-verify)`, which never equalled `--no-verify`, so the payload passed even where the region was already treated as code. Trailing quotes, brackets and separators are now trimmed before the flag comparison.

Worth recording: hole 2 is not new. Running `checkCommand` from `origin/main` directly, these are all **unblocked on main**:

```
false  node -e 'require("child_process").execSync("git push --no-verify")'
false  python3 -c "print('git commit --no-verify')"
false  python3 -c "import subprocess; subprocess.run(['git', 'commit', '--no-verify'])"
```

So the argv0-only change narrowed one shape (`python3 -c "git commit --no-verify"`, bare and space separated) that main did catch, and the branch now blocks strictly more than main in this area.

**Verification.** `tests/hooks/block-no-verify.test.js`: 43 pass, 0 fail, two tests added. Blocked: the ten evaluator payloads above, including nested `execSync` and `os.system` forms. Allowed: script-argument and printing forms, plus `node -e "console.log('use --no-verify only in emergencies')"`. Negative control: reverting the evaluator check alone fails the new test (42/1), and reverting the punctuation trim alone fails it too (42/1), so both halves are load bearing. `tests/hooks/cursor-block-no-verify.test.js` 14/14 and `tests/hooks/bash-hook-dispatcher.test.js` 6/6 stay green, and eslint is clean on the three files.

**Known residual, not addressed here.** A git invocation written as a quoted argv list inside a code payload is still not detected:

```
python3 -c "import subprocess; subprocess.run(['git', 'commit', '--no-verify'])"
```

`main` does not block it either. Detecting it means parsing each runtime's own syntax rather than shell words, so I left it out instead of half-covering it, and dropped it from the test list rather than asserting a behaviour the hook does not have.

File: scripts/hooks/block-no-verify.js [hunk 11]
```diff
@@ -289,6 +342,76 @@ function isInComment(input, idx) {
...
+  'osascript',
+  'tclsh',
+  'groovy',
+  'julia',
+  'elixir',
+  'erl',
+]);
+
+/** `-e`, `-E`, `--eval`, `-c`, `-p`, `--print`, `-r` and deno's `eval`. */
+const EVAL_FLAG = /(?:^|\s)(?:-{1,2}(?:e|eval|c|command|p|print|r)|eval)(?:=|\s|$)/i;
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Correct, and fixed in 5e6de0d. One shared expression cannot carry every runtime's flags: `-r` preloads a module for node and is the eval flag for php, so the letter alone decides nothing.

Each runtime now carries its own list:

```js
const CODE_EVALUATORS = new Map([
  ['node', ['-e', '--eval', '-p', '--print']],
  ['deno', ['-e', '--eval', 'eval']],
  ['python3', ['-c']],
  ['perl', ['-e', '-E']],
  ['php', ['-r']],
  ...
]);
```

A flag matches as a whole word or as its `=` form, so a script path that happens to end in one of them does not qualify either.

**Verification.** `tests/hooks/block-no-verify.test.js` is 45 pass, 0 fail, with two tests added:

- allowed: `node -r setup.js cli.js 'git push --no-verify'`, `node --require ts-node/register cli.js "git commit --no-verify -m x"`, `ruby -r./setup cli.rb '...'`, `python3 -B check.py '...'`
- blocked: `node -p`, `php -r`, `perl -E`, `lua -e` payloads that spawn git

Negative control: adding `-r` back to node's list fails the first of those two tests. `tests/hooks/cursor-block-no-verify.test.js` stays 14/14 and eslint is clean on both files.

**Comment by chenhz01** [comment]
Only human note on this thread is the rebase, so one narrow gap from reading the new scanner: `awk`/`gawk` fall through both sets — their first quoted argument is program source with no eval flag (`awk 'BEGIN{system("git commit --no-verify")}'`), so `evaluatesQuotedArgument` returns false and the quoted git is treated as data. Adding awk/gawk/mawk to CODE_EVALUATORS under an any-argument rule (program source, flag or not) closes it. Related nit: the wrapper/evaluator lists will keep growing one escape at a time (`expect -c` is next in line); a test asserting the deny path for unknown argv0 carrying git-shaped payloads would at least pin the intended semantics as the lists evolve. If you want this implemented rather than advised, I take fixed-scope contracts on repos like this — feel free to reach out.

**Comment by L4XB** [comment]
Thanks, fixed in 599a14fd:

- `awk`, `gawk`, `mawk` and `nawk` now keep every quoted argument subject to the guard, since their first operand is program source with no eval flag. The trade-off is that an awk pattern that only mentions a bypass flag is checked too. `expect` joins the evaluators through `-c`, and an `expect` script path with a bypass in its arguments stays allowed.
- Testing awk turned up a gap in this PR itself. When a payload went on after the call that runs git, as in `system("git push --no-verify") }` or `execSync("git push --no-verify"); console.log("done")`, the closing `"` read to the tokenizer as an opening quote. The rest of the payload then landed in the flag's token, so the flag was not recognised. `flagToken` now cuts the token at its first closing punctuation or whitespace instead of trimming only the end.

`block-no-verify.test.js` passes 47 of 47, and so do the other test files covering the hook and `shell-quotes`. Each change fails a test when reverted.

File: scripts/hooks/block-no-verify.js [hunk 12]
```diff
@@ -289,6 +342,89 @@ function isInComment(input, idx) {
...
+ * be inspected for bypass flags. A double-quoted string that contains a
+ * command substitution (`"$(git ...)"`, "`git ...`") runs git before any
+ * program receives it, so it is never data.
+ */
+function isQuotedDataArgument(input, idx) {
+  const region = quotedRegionAt(input, idx);
+  if (region === null || region.argv0 === '') return false;
+  if (region.substitution) return false;
+  const base = commandBasename(region.argv0);
+  if (base === 'git' || COMMAND_WRAPPERS.has(base) || SOURCE_OPERAND_RUNTIMES.has(base)) return false;
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Valid, and a regression from this PR: `main` checked every quoted `git`, so it blocked this. Fixed in f129cbb0. `find`, `fd`/`fdfind` and `parallel` join the command wrappers, because each runs a command line of its own for every file or input line. New cases cover `find -exec sh -c`, `find -execdir bash -c`, `fd -x sh -c` and `parallel`, and removing any one of the three names fails its case. 48/48 pass.

File: scripts/hooks/block-no-verify.js [hunk 13]
```diff
@@ -35,6 +37,58 @@ const GIT_COMMANDS_WITH_NO_VERIFY = [
...
+  'unshare',
+  'chroot',
+  'runuser',
+  'npx',
+  'bunx',
+  'pnpx',
+  // Run a command for each file or input line: find -exec, fd -x, parallel.
+  'find',
+  'fd',
+  'fdfind',
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Agreed, fixed in e8574210. `find`, `fd` and `fdfind` moved from the command wrappers to `CODE_EVALUATORS` with the flags that launch a command: `-exec`/`-execdir`/`-ok`/`-okdir` for find, and `-x`/`--exec`/`-X`/`--exec-batch` for fd. A quoted operand counts as code only when one of those flags comes before it, the same rule `node -e` follows. `find . -name 'git commit --no-verify'`, a `-printf` format and a plain `fd` pattern are allowed again, and the `-exec`/`-execdir`/`-x` cases stay blocked. `parallel` stays a wrapper. 49/49 pass. Dropping `-exec`, dropping `-execdir`, dropping the fd entry, or putting find back as a wrapper each fails a test.

File: scripts/hooks/block-no-verify.js [hunk 14]
```diff
@@ -289,6 +345,93 @@ function isInComment(input, idx) {
...
+ * Whether the words between the program and its quoted argument turn that
+ * argument into code. A flag is matched as a whole word, or as the `=` form
+ * of itself, so a path that happens to end in one of them cannot qualify.
+ */
+function evaluatesQuotedArgument(input, region, base) {
+  const flags = CODE_EVALUATORS.get(base);
+  if (flags === undefined) return false;
+  const words = input.slice(region.argv0Start, region.start).split(/\s+/);
+  return words.some((word) => flags.some((flag) => word === flag || word.startsWith(`${flag}=`)));
+}
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/block-no-verify.js]: Valid, fixed in cba56e96. After an exec flag, the launched program now decides, as it would at the top level. `find . -exec grep 'git push --no-verify' {} \;` is allowed, and `-exec sh -c '…' \;` is still checked. After `\;`, `';'` or `{} +`, the arguments are find's own again. A launched program named by an expansion (`-exec $SHELL -c …`) stays checked.

While testing this I found more forms that `main` blocks and this branch let through: `echo '…' | sh`, `$(command -v sh) -c '…'`, `$(echo '…')`, `source <(echo '…')`, `builtin eval`, `trap '…' EXIT` and `noglob sh -c`. 5fe77751 closes them. A quoted string is checked when a later pipe feeds a shell, a wrapper, a runtime without a script, or an expanded program name; when its statement opens inside or right after a substitution; and when `builtin`, `noglob`, `nocorrect`, `trap` or `alias` receives it. The description now has a table of what counts as data and what is checked.
